File: capacitor-plugins/capacitor-native/ios/Sources/AonsokuNativePlugin/Audio/NativeShuffleEngine.swift
import Foundation

class NativeShuffleEngine {
    func shuffleWithGapAvoidance(_ songs: [QueueSong], history: [String]) -> [QueueSong] {
        if history.isEmpty {
            var result = songs
            fisherYatesInPlace(&result)
            return result
        }

        let historyIndex = Dictionary(
            history.enumerated().map { ($1, $0) },
            uniquingKeysWith: { _, last in last }
        )

        var fresh: [QueueSong] = []
        var recent: [QueueSong] = []

        for song in songs {
            if historyIndex[song.id] != nil {
                recent.append(song)
            } else {
                fresh.append(song)
            }
        }

        fisherYatesInPlace(&fresh)
        recent.sort { (historyIndex[$0.id] ?? 0) < (historyIndex[$1.id] ?? 0) }

        return fresh + recent
    }

    func pickRandomStartIndex(
        count: Int,
        startHistory: [String],
        getId: (Int) -> String
    ) -> Int {
        guard count > 0 else { return 0 }

        let historySet = Set(startHistory)
        let maxAttempts = min(count, startHistory.count + 10)

        for _ in 0..<maxAttempts {
            let idx = Int.random(in: 0..<count)
            if !historySet.contains(getId(idx)) { return idx }
        }

        for i in 0..<count {
            if !historySet.contains(getId(i)) { return i }
        }

        return Int.random(in: 0..<count)
    }

    func pushToHistory(_ history: [String], id: String, maxLen: Int) -> [String] {
        var result = history.filter { $0 != id }
        result.append(id)
        if result.count > maxLen {
            result = Array(result.suffix(maxLen))
        }
        return result
    }

    private func fisherYatesInPlace<T>(_ array: inout [T]) {
        guard array.count > 1 else { return }
        for i in stride(from: array.count - 1, through: 1, by: -1) {
            let j = Int.random(in: 0...i)
            array.swapAt(i, j)
        }
    }
}

File: capacitor-plugins/capacitor-native/android/src/main/java/github/realtvop/aonsoku/plugins/audio/NativeShuffleEngine.kt
package github.realtvop.aonsoku.plugins.audio

import kotlin.random.Random

class NativeShuffleEngine {
    fun shuffleWithGapAvoidance(songs: List<QueueSong>, history: List<String>): List<QueueSong> {
        if (history.isEmpty()) {
            val result = songs.toMutableList()
            fisherYatesInPlace(result)
            return result
        }

        val historyIndex = history.mapIndexed { idx, id -> id to idx }.toMap()

        val fresh = mutableListOf<QueueSong>()
        val recent = mutableListOf<QueueSong>()

        for (song in songs) {
            if (historyIndex.containsKey(song.id)) {
                recent.add(song)
            } else {
                fresh.add(song)
            }
        }

        fisherYatesInPlace(fresh)
        recent.sortBy { historyIndex[it.id] ?: 0 }

        return fresh + recent
    }

    fun pickRandomStartIndex(
        count: Int,
        startHistory: List<String>,
        getId: (Int) -> String
    ): Int {
        if (count <= 0) return 0

        val historySet = startHistory.toSet()
        val maxAttempts = minOf(count, startHistory.size + 10)

        for (i in 0 until maxAttempts) {
            val idx = Random.nextInt(count)
            if (!historySet.contains(getId(idx))) return idx
        }

        for (i in 0 until count) {
            if (!historySet.contains(getId(i))) return i
        }

        return Random.nextInt(count)
    }

    fun pushToHistory(history: List<String>, id: String, maxLen: Int): List<String> {
        val result = history.filter { it != id }.toMutableList()
        result.add(id)
        if (result.size > maxLen) {
            return result.subList(result.size - maxLen, result.size)
        }
        return result
    }

    private fun <T> fisherYatesInPlace(list: MutableList<T>) {
        if (list.size <= 1) return
        for (i in list.size - 1 downTo 1) {
            val j = Random.nextInt(i + 1)
            val temp = list[i]
            list[i] = list[j]
            list[j] = temp
        }
    }
}
